File: ui-preview/src/main/java/l/files/ui/preview/Preview.java
package l.files.ui.preview;

import android.content.Context;
import android.graphics.Bitmap;
import android.util.DisplayMetrics;

import java.io.IOException;
import java.util.concurrent.Future;

import javax.annotation.Nullable;

import l.files.fs.Path;
import l.files.fs.Stat;
import l.files.ui.base.graphics.Rect;
import l.files.ui.base.graphics.ScaledBitmap;

import static java.lang.Boolean.TRUE;
import static l.files.base.Objects.requireNonNull;

public final class Preview {

    @Nullable
    private static Preview instance;

    public static Preview get(Context context) {
        synchronized (Preview.class) {
            if (instance == null) {
                java.io.File dir = context.getExternalCacheDir();
                if (dir == null) {
                    dir = context.getCacheDir();
                }
                Path cacheDir = Path.create(dir);
                instance = new Preview(context.getApplicationContext(), cacheDir);
                instance.cleanupAsync();
            }
        }
        return instance;
    }

    private final PersistenceCache<Rect> sizeCache;
    private final PersistenceCache<String> mediaTypeCache;
    private final PersistenceCache<Boolean> noPreviewCache;
    private final ThumbnailMemCache blurredThumbnailMemCache;
    private final ThumbnailMemCache thumbnailMemCache;
    private final ThumbnailDiskCache thumbnailDiskCache;

    final DisplayMetrics displayMetrics;
    final Path cacheDir;
    final Context context;

    Preview(Context context, Path cacheDir) {
        this.context = requireNonNull(context);
        this.cacheDir = requireNonNull(cacheDir);
        this.displayMetrics = requireNonNull(context).getResources().getDisplayMetrics();
        this.sizeCache = new RectCache(cacheDir);
        this.mediaTypeCache = new MediaTypeCache(cacheDir);
        this.noPreviewCache = new NoPreviewCache(cacheDir);
        // TODO refactor so that size here and size in decoders are in same place
        this.blurredThumbnailMemCache = new ThumbnailMemCache(context, false, 0.05f);
        this.thumbnailMemCache = new ThumbnailMemCache(context, true, 0.20f);
        this.thumbnailDiskCache = new ThumbnailDiskCache(cacheDir);
    }

    public void writeCacheAsyncIfNeeded() {
        sizeCache.writeAsyncIfNeeded();
        mediaTypeCache.writeAsyncIfNeeded();
        noPreviewCache.writeAsyncIfNeeded();
    }

    public void readCacheAsyncIfNeeded() {
        sizeCache.readAsyncIfNeeded();
        mediaTypeCache.readAsyncIfNeeded();
        noPreviewCache.readAsyncIfNeeded();
    }

    private void cleanupAsync() {
        thumbnailDiskCache.cleanupAsync();
    }

    @Nullable
    public Bitmap getBlurredThumbnail(Path path, Stat stat, Rect constraint, boolean matchTime) {
        return blurredThumbnailMemCache.get(path, stat, constraint, matchTime);
    }

    void putBlurredThumbnail(Path path, Stat stat, Rect constraint, Bitmap thumbnail) {
        blurredThumbnailMemCache.put(path, stat, constraint, thumbnail);
    }

    @Nullable
    public Bitmap getThumbnail(Path path, Stat stat, Rect constraint, boolean matchTime) {
        return thumbnailMemCache.get(path, stat, constraint, matchTime);
    }

    void putThumbnail(Path path, Stat stat, Rect constraint, Bitmap thumbnail) {
        thumbnailMemCache.put(path, stat, constraint, thumbnail);
    }

    @Nullable
    ScaledBitmap getThumbnailFromDisk(
            Path path,
            Stat stat,
            Rect constraint,
            boolean matchTime) throws IOException {
        return thumbnailDiskCache.get(path, stat, constraint, matchTime);
    }

    Future<?> putThumbnailToDiskAsync(
            Path path,
            Stat stat,
            Rect constraint,
            ScaledBitmap thumbnail) {
        return thumbnailDiskCache.putAsync(path, stat, constraint, thumbnail);
    }

    @Nullable
    public Rect getSize(Path path, Stat stat, Rect constraint, boolean matchTime) {
        return sizeCache.get(path, stat, constraint, matchTime);
    }

    void putSize(Path path, Stat stat, Rect constraint, Rect size) {
        sizeCache.put(path, stat, constraint, size);
    }

    @Nullable
    String getMediaType(Path path, Stat stat, Rect constraint, boolean matchTime) {
        return mediaTypeCache.get(path, stat, constraint, matchTime);
    }

    void putMediaType(Path path, Stat stat, Rect constraint, String media) {
        mediaTypeCache.put(path, stat, constraint, media);
    }

    public NoPreview getNoPreviewReason(Path path, Stat stat, Rect constraint) {

        if (!stat.isRegularFile()) {
            return NoPreview.NOT_REGULAR_FILE;
        }

        try {
            if (!path.isReadable()) {
                return NoPreview.FILE_UNREADABLE;
            }
        } catch (IOException e) {
            return new NoPreview(e);
        }

        if (TRUE.equals(noPreviewCache.get(path, stat, constraint, true))) {
            return NoPreview.IN_NO_PREVIEW_CACHE;
        }

        return null;
    }

    public boolean isPreviewable(Path path, Stat stat, Rect constraint) {
        return getNoPreviewReason(path, stat, constraint) == null;
    }

    void putPreviewable(Path path, Stat stat, Rect constraint, boolean previewable) {
        if (previewable) {
            noPreviewCache.remove(path, stat, constraint);
        } else {
            noPreviewCache.put(path, stat, constraint, true);
        }
    }

    @Nullable
    public Decode get(
            Path path,
            Stat stat,
            Rect constraint,
            Callback callback) {

        if (!isPreviewable(path, stat, constraint)) {
            return null;
        }

        return new Decode(path, stat, constraint, callback, this)
                .executeOnPreferredExecutor();
    }

    public void clearThumbnailCache() {
        thumbnailMemCache.clear();
    }

    public void clearBlurredThumbnailCache() {
        blurredThumbnailMemCache.clear();
    }

    public interface Callback {

        void onPreviewAvailable(Path path, Stat stat, Bitmap thumbnail);

        void onBlurredThumbnailAvailable(Path path, Stat stat, Bitmap thumbnail);

        void onPreviewFailed(Path path, Stat stat, Object cause);

    }

}

File: ui-preview/src/main/java/l/files/ui/preview/Preview.kt
package l.files.ui.preview

import android.content.Context
import android.graphics.Bitmap
import androidx.annotation.MainThread
import l.files.ui.base.graphics.Rect
import l.files.ui.base.graphics.ScaledBitmap
import java.io.IOException
import java.nio.file.Files.isReadable
import java.nio.file.Files.isRegularFile
import java.nio.file.Path
import java.time.Instant
import java.util.concurrent.Future

internal data class BlurredThumbnail(val bitmap: Bitmap)

internal data class NoPreview(val cause: Any) {
  companion object {
    val FILE_UNREADABLE = NoPreview("file is unreadable")
    val NOT_REGULAR_FILE = NoPreview("not a regular file")
    val PATH_IN_CACHE_DIR = NoPreview("path is in cache directory")
    val IN_NO_PREVIEW_CACHE = NoPreview("file marked in no preview cache")
    val DECODE_RETURNED_NULL = NoPreview("decode returned null")
  }
}

class Preview internal constructor(
  context: Context,
  getCacheDir: () -> Path
) {

  internal val cacheDir by lazy { getCacheDir() }
  private val sizeCache = RectCache { cacheDir }
  private val mediaTypeCache = MediaTypeCache { cacheDir }
  private val noPreviewCache = NoPreviewCache { cacheDir }
  private val thumbnailMemCache = ThumbnailMemCache(context, true, 0.20f)
  private val thumbnailDiskCache = ThumbnailDiskCache { cacheDir }

  // TODO refactor so that size here and size in decoders are in same place
  private val blurredThumbnailMemCache =
    ThumbnailMemCache(context, false, 0.05f)

  fun writeCacheAsyncIfNeeded() {
    sizeCache.writeAsyncIfNeeded()
    mediaTypeCache.writeAsyncIfNeeded()
    noPreviewCache.writeAsyncIfNeeded()
  }

  fun readCacheAsyncIfNeeded() {
    sizeCache.readAsyncIfNeeded()
    mediaTypeCache.readAsyncIfNeeded()
    noPreviewCache.readAsyncIfNeeded()
  }

  internal fun cleanupAsync() {
    thumbnailDiskCache.cleanupAsync()
  }

  fun getBlurredThumbnail(
    path: Path,
    time: Instant,
    constraint: Rect,
    matchTime: Boolean
  ) = blurredThumbnailMemCache.get(path, time, constraint, matchTime)

  fun putBlurredThumbnail(
    path: Path,
    time: Instant,
    constraint: Rect,
    thumbnail: Bitmap
  ) {
    blurredThumbnailMemCache.put(path, time, constraint, thumbnail)
  }

  fun getThumbnail(
    path: Path,
    time: Instant,
    constraint: Rect,
    matchTime: Boolean
  ) = thumbnailMemCache.get(path, time, constraint, matchTime)

  fun putThumbnail(
    path: Path,
    time: Instant,
    constraint: Rect,
    thumbnail: Bitmap
  ) {
    thumbnailMemCache.put(path, time, constraint, thumbnail)
  }

  @Throws(IOException::class)
  fun getThumbnailFromDisk(
    path: Path,
    time: Instant,
    constraint: Rect
  ): ScaledBitmap? = thumbnailDiskCache.get(path, time, constraint, true)

  fun putThumbnailToDiskAsync(
    path: Path,
    time: Instant,
    constraint: Rect,
    thumbnail: ScaledBitmap
  ): Future<*> = thumbnailDiskCache.putAsync(path, time, constraint, thumbnail)

  fun getSize(
    path: Path,
    time: Instant,
    constraint: Rect,
    matchTime: Boolean
  ): Rect? = sizeCache.get(path, time, constraint, matchTime)

  fun putSize(
    path: Path,
    time: Instant,
    constraint: Rect,
    size: Rect
  ) {
    sizeCache.put(path, time, constraint, size)
  }

  fun getMediaType(
    path: Path,
    time: Instant,
    constraint: Rect
  ): String? = mediaTypeCache.get(path, time, constraint, true)

  fun putMediaType(
    path: Path,
    time: Instant,
    constraint: Rect,
    media: String
  ) {
    mediaTypeCache.put(path, time, constraint, media)
  }

  internal fun getNoPreviewReason(
    path: Path,
    time: Instant,
    constraint: Rect
  ): NoPreview? {
    if (!isRegularFile(path)) {
      return NoPreview.NOT_REGULAR_FILE
    }
    if (!isReadable(path)) {
      return NoPreview.FILE_UNREADABLE
    }
    val noPreview = noPreviewCache.get(path, time, constraint, true)
    return if (java.lang.Boolean.TRUE == noPreview) {
      NoPreview.IN_NO_PREVIEW_CACHE
    } else {
      null
    }
  }

  fun isPreviewable(
    path: Path,
    time: Instant,
    constraint: Rect
  ): Boolean {
    return getNoPreviewReason(path, time, constraint) == null
  }

  fun putPreviewable(
    path: Path,
    time: Instant,
    constraint: Rect,
    previewable: Boolean
  ) {
    if (previewable) {
      noPreviewCache.remove(path, constraint)
    } else {
      noPreviewCache.put(path, time, constraint, true)
    }
  }

  fun get(
    path: Path,
    time: Instant,
    constraint: Rect,
    callback: Callback,
    context: Context
  ): Decode? = if (!isPreviewable(path, time, constraint)) {
    null
  } else {
    Decode(path, time, constraint, callback, this)
      .executeOnPreferredExecutor(context)
  }

  fun clearThumbnailCache() {
    thumbnailMemCache.clear()
  }

  fun clearBlurredThumbnailCache() {
    blurredThumbnailMemCache.clear()
  }

  interface Callback {
    fun onPreviewAvailable(
      path: Path,
      time: Instant,
      thumbnail: Bitmap
    )

    fun onBlurredThumbnailAvailable(
      path: Path,
      time: Instant,
      thumbnail: Bitmap
    )

    fun onPreviewFailed(path: Path, time: Instant, cause: Any)
  }

}

private var instance: Preview? = null

@MainThread
fun Context.getPreview(): Preview {
  if (instance == null) {
    instance = Preview(applicationContext) {
      (externalCacheDir ?: cacheDir).toPath()
    }
    instance!!.cleanupAsync()
  }
  return instance!!
}
